### DISCUSSION TITLE
Write an ElectionGuard Verifier

### DISCUSSION THREAD

**Discussion by keithrfung** [microsoft/electionguard#107] [General]
# Writing a Verifier

To write a basic verifier there are two main parts:
**1.	Ballot Verifier
2.	Decryption Verifier**

To build a **_complete_** verifier consists of verifying the integrity of every ballot cast in an election and verifying that the aggregate ballot (formed by combining each of he individual cast ballots) and that all of the spoiled ballots have been correctly decrypted.

The steps provided by Dr. Josh Benaloh himself are listed here:
📝  [Building a Verifier](https://github.com/microsoft/electionguard/files/5298914/EG.Verifier.Subtasks.docx)
📝  [Verifier Construction](https://github.com/microsoft/electionguard/files/5278086/EG.Verifier.Construction.docx)

This example verifier was completed by a set of Josh Benaloh's students at University of Washington:
📝  [Example Verifier by Rainbow Huang](https://github.com/rainbowhuanguw/ElectionGuard-verifier-python)

## The Ask

We'd love to see _any_ part of a verifier. 
1. Grab our [sample data](https://github.com/microsoft/electionguard-python/tree/main/data/5-ballots)
2. Code your verifier 
**Note:** Any language and any verifier counts. Verify just a single step or write a Ballot Verifier, Decryption Verifier, or complete Election Verifier.
3. [Push it to GitHub](https://docs.github.com/en/github/importing-your-projects-to-github/adding-an-existing-project-to-github-using-the-command-line)
4. [Create a PR](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) to add your name and repo to the verifier list.

## Ballot Verifier
_**Note:** All steps reference [Verifier Construction Document](https://github.com/microsoft/electionguard/files/5278086/EG.Verifier.Construction.docx)_

Ballot verification consists of steps 3 and 4.
- **Step 3** serves to verify that each encryption (corresponding to a selection made by a voter) is either an encryption of one (indicating that the voter selected this option) or an encryption of zero (indicating that the voter did not select this option).  When the ballots are aggregated, each component will consist of an encryption of the number of times that option was selected – i.e., the tally.  A malicious voting device could – possibly in collusion with a willing voter – encrypt a value other than one or zero and thereby compromise the integrity of the tally.  Successful verification of the step 3 equations ensures that ballots are correctly formed.
- **Step 4** serves to verify that the number of options selected for a contest to not exceed a pre-assigned limit (usually one).  For instance, in most elections, a ballot should not be allowed to cast a vote for more than one option.  This step can be omitted for yes/no referenda (voting both yes and no is equivalent to voting for neither) and for approval votes in which voters are permitted to vote for as many options s they wish.

## Decryption Verifier
_**Note:** All steps reference [Verifier Construction Document](https://github.com/microsoft/electionguard/files/5278086/EG.Verifier.Construction.docx)_

A decryption verifier serves to verify the correct decryption of a ballot and is normally applied to either an individual spoiled ballot (consisting entirely of encryptions of ones and zeros) or an aggregate ballot (consisting of encryptions of the election tallies). The decryption verification consists of steps 6 through 9 _– although steps 7 and 8 can be omitted when all guardians have completed their election tasks_.

- **Step 6** begins with two equations that confirm that all individual cast ballots have been correctly aggregated.  (This is omitted when applied to a single spoil ballot.)  The remaining equations verify that the partial decryptions done by each guardian are correct.
- **Steps 7 and 8 _(Optional)_** confirm that any decryption data which one or more guardians has failed to provide have been correctly substituted for by a quorum of other guardians.
- **Step 9** confirms that the partial decryptions provided by the guardians have been correctly combined to form a full decryption.

## Above and Beyond 
_**Note:** All steps reference [Verifier Construction Document](https://github.com/microsoft/electionguard/files/5278086/EG.Verifier.Construction.docx)_

If you want more verification goodness, there are a few remaining steps to full verification of an election that are described here for completeness. 

- **Step 1** serves to verify the suitability of the parameters used in an election.
- **Step 2** serves to verify the integrity of the public keys published by the election guardians.  Although public verification of this step is desirable, it would suffice to have the guardians simply take responsibility for checking each other.
- **Step 5** serves to verify that the ballots in an election have been correctly linked.  This makes it more difficult for an election administrator to simply discard ballots in hopes that no one will notice.
- **Step 10** is implicitly included within the description of the decryption verifier.

## Stuck?

Post questions here or send them to electionguard@microsoft.com

**Comment by rexdivakar** [comment]
Hey @keithrfung  I would like to work on this can u please assign it to me

**Comment by keithrfung** [comment]
@rexdivakar This is an open issue for anyone to code for Hacktoberfest. No need for assignment.

**Comment by JohnLCaron** [comment]
Im considering the feasibility of writing a Verifier in Java. Just starting to look at the python code, and Im wondering about the availability of the C++ library and source?

**Comment by kiniry** [comment]
The original implementation is written in C and is available via
https://github.com/microsoft/electionguard-c

Additional related artifacts and expertise are available via Free & Fair.

**Comment by JohnLCaron** [comment]
I have version 0.1 of a complete port of electionguard-python at https://github.com/JohnLCaron/electionguard-java.
Ill be looking at "writing a verifier" soon. 

Wanted to start letting people know about it; I'm open to feedback etc.

**Comment by rc-ms** [comment]
Hello @JohnLCaron thank you so much for your contribution 🎉!  My apologies for the delay in getting back. Thank you for correctly [porting from the ElectionGuard python repo](https://github.com/microsoft/electionguard-python). If you could DM an email address to correspond we would like to arrange some swag but also have some questions we would like to ask based on your experience.   (@kiniry please note the ElectionGuard-C library has been deprecated and will soon be archived).  

If you are serious about working on a verifier, please consider [adding it here](https://github.com/microsoft/electionguard/blob/main/docs/guide/Implementations.md#community-verifiers), and if you have any further questions we have [started a discussions board](https://github.com/microsoft/electionguard/discussions).

Thanks again for your work and interest.

**Comment by JohnLCaron** [comment]
Hey @rc-ms feel free to use my github profile address: jcaron1129 at gmail.
